File: tunnelcave_sandbox_web/lib/frame.ts
import { angleBetween, cross, dot, length, normalize, scale, Vec3 } from "./vector";

export interface OrthonormalFrame {
  forward: Vec3;
  right: Vec3;
  up: Vec3;
}

export function createInitialFrame(direction: Vec3): OrthonormalFrame {
  const forward = normalize(direction);
  const hint: Vec3 = Math.abs(forward[1]) < 0.9 ? [0, 1, 0] : [1, 0, 0];
  let right = normalize(cross(hint, forward));
  if (length(right) === 0) {
    right = [1, 0, 0];
  }
  const up = normalize(cross(forward, right));
  return { forward, right, up };
}

export function transportFrame(
  frame: OrthonormalFrame,
  newForward: Vec3
): OrthonormalFrame {
  const fOld = frame.forward;
  const fNew = normalize(newForward);
  const angle = angleBetween(fOld, fNew);
  if (angle < 1e-5) {
    return { forward: fNew, right: frame.right, up: frame.up };
  }
  const axis = cross(fOld, fNew);
  const axisLen = length(axis);
  if (axisLen === 0) {
    return { forward: fNew, right: frame.right, up: frame.up };
  }
  const k = [axis[0] / axisLen, axis[1] / axisLen, axis[2] / axisLen] as Vec3;
  const cos = Math.cos(angle);
  const sin = Math.sin(angle);

  const rotate = (v: Vec3): Vec3 => {
    const term1 = scale(v, cos);
    const term2 = scale(cross(k, v), sin);
    const term3 = scale(k, dot(k, v) * (1 - cos));
    return [term1[0] + term2[0] + term3[0], term1[1] + term2[1] + term3[1], term1[2] + term2[2] + term3[2]];
  };

  const right = normalize(rotate(frame.right));
  const up = normalize(rotate(frame.up));
  return { forward: fNew, right, up };
}

File: tunnelcave_sandbox/frame.py
"""Orthonormal frame utilities for the cave path."""
from __future__ import annotations

import math
from dataclasses import dataclass

from .vector import Vector3, orthonormalize


@dataclass(frozen=True)
class OrthonormalFrame:
    """Coordinate frame that travels along the tunnel path."""

    origin: Vector3
    forward: Vector3
    right: Vector3
    up: Vector3

    def transport(self, new_origin: Vector3, new_forward: Vector3) -> "OrthonormalFrame":
        """Move the frame using parallel transport.

        The method keeps the in-plane axes free of unnecessary spins by
        rotating them only by the minimal rotation that maps the old
        forward direction into the new forward direction.
        """

        old_forward = self.forward
        new_forward_norm = new_forward.normalized()
        dot = max(-1.0, min(1.0, old_forward.dot(new_forward_norm)))
        angle = math.acos(dot)
        if angle < 1e-6:
            return OrthonormalFrame(new_origin, new_forward_norm, self.right, self.up)

        axis = old_forward.cross(new_forward_norm)
        axis_length = axis.length()
        if axis_length < 1e-8:
            # The directions are opposite. We rotate around a stable axis
            # that depends only on the forward vector to avoid sudden flips.
            axis = self.right if abs(old_forward.dot(self.right)) < 0.9 else self.up
            axis_length = axis.length()
        axis = axis / axis_length

        right_rotated = _rotate_vector(self.right, axis, angle)
        up_rotated = _rotate_vector(self.up, axis, angle)
        return OrthonormalFrame(new_origin, new_forward_norm, right_rotated, up_rotated)

    @staticmethod
    def initial(origin: Vector3, forward: Vector3) -> "OrthonormalFrame":
        fwd, up, right = orthonormalize(forward, Vector3(0.0, 1.0, 0.0))
        return OrthonormalFrame(origin, fwd, right, up)


def _rotate_vector(vector: Vector3, axis: Vector3, angle: float) -> Vector3:
    """Rotate ``vector`` around ``axis`` by ``angle`` radians."""

    sin_theta = math.sin(angle)
    cos_theta = math.cos(angle)
    axis = axis.normalized()
    return (
        vector * cos_theta
        + axis.cross(vector) * sin_theta
        + axis * axis.dot(vector) * (1.0 - cos_theta)
    )
